geo/src/algorithm/closest_point.rs

```diff
@@ -121,6 +122,18 @@ impl<F: GeoFloat> ClosestPoint<F> for Polygon<F> {
     }
 }
 
+impl<F: GeoFloat> ClosestPoint<F> for Triangle<F> {
+    fn closest_point(&self, p: &Point<F>) -> Closest<F> {
+        self.to_polygon().closest_point(p)
```

Comment:
This looks like it would behave correctly, but a minor improvement request:

Creating a polygon here, will allocate a Vec (two of them actually), please instead do something like: 

`closest_of(self.coords_iter())` which will use the [CoordsIter](https://docs.rs/geo/0.18.0/geo/algorithm/coords_iter/trait.CoordsIter.html) trait without necessarily allocating on the heap.

Reply:
Thanks for the feedback @michaelkirk 
I tried this, but there's no `ClosestPoint` method for the type `Coordinate`. And it looks like `Coordinate` will need to be converted to `Point` to implement that method, in one way or another. Would that also cause a heap allocation?

Reply:
Ah right, good point. Converting  a `Coordinate` to a `Point` will not require a heap allocation.

Reply:
It's possible that this will be a larger change. I'd also be OK with merging the performance improvement in a followup since it won't affect the API. Up to you!

Reply:
Thanks Michael.
I've added a `ClosestPoint` for `Coordinate`, switched `Triangle` and `Rect` to use `self.iter()` and updated the `CHANGES.md`.

Reply:
I'm new to Rust, so I'm not sure if using `*self` for the `Coordinate` to `Point` transform will cause permanent type changes. Is this way of doing it alright?

Reply:
> 've added a ClosestPoint for Coordinate, switched Triangle and Rect to use self.iter() and updated the CHANGES.md.

The change log and the new Coordinate implementation look good, thank you!

But there's a problem of my own making: we're trying to find the point nearest to a Rect or Triangle by using only its corners. This will give incorrect results whenever the nearest point is mid line. Sorry for suggesting you go down that path, that was my fault.

So two alternative ways forward I see:
1. revert to using to_poly as you originally wanted.
2. create an implementation for Rect/Tri  that builds Lines and uses the `impl ClosestPoint for Line` to determine the closest point - this looks like it should be correct for endpoints and midline points.

Alternative 1 is simpler to write, but alternative 2 will avoid the heap allocations (and should be correct 🙈).

Also, it might be worthwhile to include a couple more test cases to catch future bad suggestions.

> I'm new to Rust, so I'm not sure if using *self for the Coordinate to Point transform will cause permanent type changes. Is this way of doing it alright?

Looks good to me. This is part of rust ownership system. `self` is a reference (you can see in the signature `&self`), but Point needs an owned `Coordinate`, not a borrowed coordinate (`&Coordinate`), so we de-reference the borrow to the owned type. Because `Coordinate` implements the `Copy` trait, rather than moving/consume `self`, a new copy of self will be made. Since copying could in theory be expensive, typically Copy is only implemented for trivial types, whereas for more complicated types (especially anything doing heap allocation) you'll have to explicitly call "thing.clone()" to get a copy.

Reply:
Ahahaha! A seemingly easy task that grows in complexity is a nice step-by-step tutorial to learning rust. 
I'll try to do it the right way, should have something next week. Thank you also for the explanation of ownership.

Reply:
I was thinking about this more and realized our current implementation `for Polygon` is incorrect. For 2D geometries, we should considering points *inside* the geometry as an Intersection.

Since it's a correctness thing, I wanted to address it right away and opened a PR here: https://github.com/georust/geo/pull/679

Reply:
Assuming that gets merged, you should merge or rebase to incorporate that fix into your work next week.

Reply:
Will do. Another question there: If we use Using only lines to determine `ClosestPoint`, wouldn't a similar interior issue affect `Triangle` and `Rect`?

Reply:
Yes exactly. You''ll have to include a similar intersection check like in #679.

Reply:
@michaelkirk I'm trying to understand: how do we determine the interior for intersection if `Triangle` and `Rect` are a set of lines joining the vertices? It's so much easier if it's converted to a polygon. What do you think?

Reply:
The intersection trait is implemented for Triangles and Rects too.